### ISSUE TITLE
Could you explain the principle behind streaming?

### ISSUE THREAD

**Issue by deeptimhe** [mosaicml/streaming#336]
It looks like streaming supporting global index in __getitem__(), so how could it accelerate data loading for large-scale sharded data? I mean, if the data index is global, the training loader must access each shards randomly. What's the implementation difference compared with webdataset?

Of course, it's an awesome project and I am learning to use it.

**Comment by knighton** [comment]
Random access is available if you need it (e.g. for data inspection), but won't be performant.

What is performant are the node-aware, shard-aware shuffling algorithms. It's an IterableDataset, so you rely on `__iter__`.

See method `_get_work` here: https://github.com/mosaicml/streaming/blob/e2d0431c35e4a2fa9435edbe0f432f6c933f75ad/streaming/base/dataset.py#L753

**Comment by deeptimhe** [comment]
So you mean, to achieve excellent performance, I need to rewrite `__iter__()` instead of `__getitem__()`?
However, I found all examples are based on `__getitem__()` (https://docs.mosaicml.com/projects/streaming/en/stable/getting_started/user_guide.html#, https://docs.mosaicml.com/projects/streaming/en/stable/examples/facesynthetics.html)

What should I do to use the node-aware, shard-aware shuffling algorithms? Do you know if there are any examples?

**Comment by knighton** [comment]
We provide both `__getitem__` and `__iter__` for you. `__iter__` is very complicated and is the reason for StreamingDataset to exist. `__getitem__` provides numpy array-style indexing, and calls `get_item` per individual sample. If you used custom data types in your streaming dataset, you would subclass StreamingDataset overriding `get_item` to do your custom data type deserialization.

You get `__iter__`'s fancy shuffling by default by doing nothing, but can configure it manually via arguments to StreamingDataset like `shuffle_algo`.

**Comment by deeptimhe** [comment]
Thanks a lot for your kind response. Do you mean I just need to override `__getitem__` and StreamingDataset do all the rest to achieve `__iter__`?
Sounds cool!

**Comment by knighton** [comment]
Yes, if you rolled your own data types.

If our collection of built-in data types was enough for your dataset, you don't have to do anything at all.

For example, here is the do-nothing ImageNet streaming dataset:
- Writing: https://github.com/mosaicml/streaming/blob/main/streaming/vision/convert/imagenet.py
- Reading: https://github.com/mosaicml/streaming/blob/main/streaming/vision/imagenet.py

**Comment by deeptimhe** [comment]
This is actually what I need! Since iterable dataset tools like webdataset are hard to learn from scratch. You make everything much simple!
I will build MDS files for my own dataset.

**Comment by knighton** [comment]
Good luck, and be sure to be very mindful about shuffling (including globally pre-shuffling before writing the shards) as convergence can be very sensitive to it.

Also there are other features that aren't really advertised like weighting and mixing datasets together on the fly (streams).

**Comment by deeptimhe** [comment]
BTW, for multiprocess dataset conversation, it may be easier to directly use torch.utils.data.Dataloader and set batch_size=1 with larger num_workers.

**Comment by knighton** [comment]
You can also convert to MDS using multiple nodes, by processing slices of the original dataset and concatenating the resulting streaming datasets afterward. A streaming dataset index file is just a listing of its shard files with essential metadata. They can be subsetted, rearranged, concatenated, etc. however you want.

```
root/
    index.json <- joint index file you create from all the parts
    part.00000/
        index.json
        shard.00000.mds
        ...
    part.00001/
        index.json
        ...
    ...
```

**Comment by karan6181** [comment]
Hi @deeptimhe , closing this issue for now. Please feel free to re-open if the issue is not addressed. Thanks!
